### ISSUE TITLE
Can't get SocketManager to work properly

### ISSUE THREAD

**Issue by WilhelmW** [Facepunch/Facepunch.Steamworks#513]
```
public class TestClient : ConnectionManager
{
	public override void OnConnecting(ConnectionInfo info)
	{
		Debug.Log("Connecting");
	}

	public override void OnConnected(ConnectionInfo info)
	{
		Debug.Log("Connected");

		string str = "Msg for Server";
		Connection.SendMessage(Marshal.StringToHGlobalUni(str), str.Length * 2 + 1, SendType.Reliable);
	}

	public override void OnDisconnected(ConnectionInfo info)
	{
		Debug.Log("Disconnected");
	}

	public override void OnMessage(IntPtr data, int size, long messageNum, long recvTime, int channel)
	{
		Debug.Log("Msg received: " + Marshal.PtrToStringUni(data));
	}
}
```

```
public class TestServer : SocketManager
{
	public override void OnConnecting(Connection connection, ConnectionInfo data)
	{
		connection.Accept();
		Debug.Log($"{data.Identity} is connecting");
	}

	public override void OnConnected(Connection connection, ConnectionInfo data)
	{
		Debug.Log($"{data.Identity} has joined the game");

		string str = "Msg for client";
		connection.SendMessage(Marshal.StringToHGlobalUni(str), str.Length * 2 + 1, SendType.Reliable);
	}

	public override void OnDisconnected(Connection connection, ConnectionInfo data)
	{
		Debug.Log($"{data.Identity} has disconnected");
	}

	public override void OnMessage(Connection connection, NetIdentity identity, IntPtr data, int size, long messageNum, long recvTime, int channel)
	{
		Debug.Log($"We got a message from {identity}!");

		connection.SendMessage(data, size, SendType.Reliable);
	}
}
```

```
public class Test : MonoBehaviour
{
	TestServer server;
	TestClient client;

	private void Awake()
	{
		SteamManager.Initialize();

		server = SteamNetworkingSockets.CreateNormalSocket<TestServer>(NetAddress.AnyIp(12345));
		client = SteamNetworkingSockets.ConnectNormal<TestClient>(NetAddress.LocalHost(12345));
	}

	private void Update()
	{
		server.Receive();
		client.Receive();
	}
}
```

Output in Unity:
Connecting
ip:[::1]:56920 is connecting
ip:[::1]:56920 has joined the game
Connected
Msg received: Msg for client⤀뚿졚솄嬄ﾠᑄ젪ꗜ訛좘쯱⠶תּ⿩姏㉉鿀흆ɔ

Problem:
For some reason the message to the server never triggers the OnMessage method in the TestServer class.
Most of the time there are random characters after the message I receive.

**Comment by WilhelmW** [comment]
I managed to resolve the problem with the random characters.
Apparently I needed two null bytes to terminate a unicode string.

**Comment by Rohansi** [comment]
You need to pump messages with `Receive()` otherwise `OnMessage` will not be called.

**Comment by WilhelmW** [comment]
I am already using Receive().
`private void Update()
{
	server.Receive();
	client.Receive();
}`

In which other way should I use it?

**Comment by bthomas2622** [comment]
@WilhelmW having same issue getting server onmessage to proc, what did you mean by adding 2 null bytes? How would one do that?

**Comment by WilhelmW** [comment]
The 2 null bytes are for terminating the unicode string I am sending. It was for fixing the problem about receiving random characters after the message. I don't think it has anything to do with the other problem you also seem to experience.

If I calculate the length with `str.Length * 2 + 1` I only send 1 null byte. If I calculate the length with `str.Length * 2 + 2` I send 2 null bytes.

**Comment by bthomas2622** [comment]
Gotcha gotcha that is helpful. @WilhelmW when you said 

> "Problem:
> For some reason the message to the server never triggers the OnMessage method in the TestServer class."

That seems like what I'm experiencing. Was that something you resolved separately to the random character issue? Or resolved itself?

**Comment by WilhelmW** [comment]
@bthomas2622 That issue is still unresolved.

**Comment by bthomas2622** [comment]
@WilhelmW got a fix! See #467
